# Homework 1: Introduction to Machine Learning

ML Zoomcamp 2026. The notebook follows the official question order and uses the pinned 2026 dataset.

## Setup

In [1]:
from pathlib import Path
import numpy as np
import pandas as pd

DATA_URL = (
    'https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/'
    'main/cohorts/2026/data/car_fuel_efficiency_2026.csv'
)

## Q1. Pandas version

In [2]:
pd.__version__

'2.2.3'

## Load the data

Read the course's pinned 2026 release directly from its authoritative source.

In [3]:
df = pd.read_csv(DATA_URL)
df.head()

   model_year  origin  ... acceleration fuel_efficiency_mpg
0        2006  Europe  ...          NaN                31.9
1        2008  Europe  ...          NaN                31.3
2        1996    Asia  ...         17.3                27.5
3        1989  Europe  ...         18.7                28.5
4        1994     USA  ...         17.5                31.0

[5 rows x 11 columns]

## Q2. Records count

In [4]:
record_count = len(df)
record_count

10000

**Answer: 10000**

## Q3. Fuel types

In [5]:
fuel_type_count = df['fuel_type'].nunique()
fuel_type_count

3

**Answer: 3**

## Q4. Missing values

In [6]:
missing_by_column = df.isna().sum()
missing_by_column[missing_by_column > 0]

horsepower      877
acceleration    264
dtype: int64

In [7]:
columns_with_missing_values = df.isna().any().sum()
columns_with_missing_values

np.int64(2)

**Answer: 2**

## Q5. Maximum fuel efficiency for Asian cars

In [8]:
asia_max_fuel_efficiency = df.loc[
    df['origin'].eq('Asia'), 'fuel_efficiency_mpg'
].max()
asia_max_fuel_efficiency

np.float64(41.2)

**Answer: 41.2**

## Q6. Median horsepower after imputing the mode

In [9]:
horsepower_median_before = df['horsepower'].median()
horsepower_mode = df['horsepower'].mode().iloc[0]
horsepower_median_after = df['horsepower'].fillna(horsepower_mode).median()

{
    'median_before': horsepower_median_before,
    'mode_used_for_imputation': horsepower_mode,
    'median_after': horsepower_median_after,
}

{'median_before': np.float64(254.0), 'mode_used_for_imputation': np.float64(252.0), 'median_after': np.float64(252.0)}

The median falls from 254 to 252 after filling missing values with the mode.

**Answer: Yes, it decreased.**

## Q7. Sum of weights

In [10]:
X = (
    df.loc[df['origin'].eq('Asia'), ['vehicle_weight', 'model_year']]
    .head(7)
    .to_numpy()
)
XTX = X.T @ X
XTX_inverse = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = XTX_inverse @ X.T @ y

{'weights': w, 'sum_of_weights': w.sum()}

{'weights': array([0.13644777, 0.2327492 ]), 'sum_of_weights': np.float64(0.36919696904925486)}

**Answer: 0.369**

## Checks

In [11]:
assert record_count == 10_000
assert fuel_type_count == 3
assert columns_with_missing_values == 2
assert asia_max_fuel_efficiency == 41.2
assert horsepower_median_after < horsepower_median_before
assert np.isclose(w.sum(), 0.36919696904925486)
print('All Homework 1 checks passed.')

All Homework 1 checks passed.
